File: LeetCode/3-longest_substring_without_repeating_characters.py
from collections import Counter

class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        count = Counter()
        maxlen = 0
        n, k = len(s), 0
        for i, c in enumerate(s):
            while count[c] > 0:
                maxlen = max(maxlen, i - k)
                count[s[k]] -= 1
                k += 1
            count[c] += 1
        maxlen = max(maxlen, n - k)
        return maxlen

File: LeetCode/3-longest_substring_without_repeating_characters.cpp
#include <string>
#include <algorithm>
#include <cassert>

class Solution {
public:
    int lengthOfLongestSubstring(const std::string& s)
    {
        int used[256];
        std::fill_n(used, 256, 0);

        const int n = s.size();
        int maxlen = 0;
        int l = 0, r = 0;
        while (r < n) {
            used[s[r]]++;
            while (used[s[r]] > 1) {
                used[s[l++]]--;
                assert(l <= r);
            }
            maxlen = std::max(maxlen, r - l + 1);
            r++;
        }

        return maxlen;
    }
};
